DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+{
+    int[] b;
+    this(ref B rhs)
+    {
+        this.b = rhs.b.dup;
+    }
+}
+
+struct C
+{
+    A a;
+    B b;
+    /* the following copy constructor is generated for C:
+
+    this(ref C rhs)
+    {
+        this.a = rhs.a;       // which, in turn, is lowered to a.__copyCtor(rhs.a);
+        this.b = rhs.b;       // which, in turn, is lowered to b.__copyCtor(rhs.b);
+    }
+    */
+}
+```
+
+#### Generating `opAssign` from a copy constructor
+
+The copy constructor is used to initialize an object from another object, whereas `opAssign` is
+used to copy an object to another object that has already been initialized:
+
+```d
+struct A
+{
+    int a;
+    immutable int id;
+    this(int a, int b)
+    {
+        this.a = a;
+        this.b = b;
+    }
+    this(ref A rhs)
+    {
+        this.a = rhs.a;
+        this.b = rhs.b;
+    }
+    void opAssign(A rhs)
+    {
+        this.a = rhs.a;
+    }
+}
+
+void main()
+{
+    A a = A(2);
+    A b = a;      // calls copy constructor;
+    b = a;        // calls opAssign;
+}
+```
+
+The reason why both the copy constructor and the `opAssign` method are needed is because
+the two are type checked differently: `opAssign` is type checked as a normal function whereas
+the copy constructor is type checked as a constructor (where the first assignment of non-mutable
+fields is allowed). However, in the majority of cases, the copy constructor body is identical
+to the `opAssign` one:
+
+```d
+struct A
+{
+    int a;
+    this(int a)
+    {
+        this.a = a;
+    }
+    this(ref A rhs)
+    {
+        this.a = rhs.a;
+    }
+    void opAssign(S rhs)
+    {
+        this.a = rhs.a;
+    }
+}
+
+void main()
+{
+    A a = A(2);
+    A b = a;      // calls copy constructor;
+    b = a;        // calls opAssign;
+}
+```
+
+In order to avoid the code duplication resulting from such situations, ideally, the user could
+define a single method that deals with both copy construction and normal copying. This is possible
+in certain situations which will be discussed below:
+
+1. If the struct contains any `const`/`immutable` fields, it is impossible to use the copy constructor
+for `opAssign`, because the copy constructor might initialize them. Even if the copy constructor doesn't
+modify the `const`/`immutable` fields the compiler has to analyze the function body to know that, which
+is problematic in situations when the body is missing. In conclusion, `opAssign` can be generated from the
+copy constructor when the `struct` contains only assignable (mutable) fields.
+
+2. The copy constructor signature is: `this(ref $q1 S rhs) $q2`, where `q1` and `q2` represent the
+qualifiers that can be applied to the function and the parameter (`const`, `immutable`, `shared`, etc.). The
+problem that arises is: depending on the values of `$q1` and `$q2` what should the signature of `opAssign` be?
+A solution might be to generate for every copy constructor its counterpart `opAssign`: `void opAssign(ref $q1 S rhs) $q2`.
```

Comment:
for every copy constructor its counterpart `opAssign`: -> the counterpart `opAssign` for each copy constructor, e.g.